### PR TITLE
Hackmud test framework

### PR DESCRIPTION
**Pull request by memetech**
Hi Soron,

I wrote up a relatively comprehensive set of scripts for running hackmud scripts outside of the game - it pre-processes the scripts to remove the #db. #s. items and and replace them with functional versions. As a result you can step through your code with Visual Studio's debugger or run in a node environment.

I dropped everything under a new directory testing/.  Please take a look and see if this is something you think people would find useful here. 

In game I am d00mlaut.

Best,

Peter

### CODE REVIEW

**Comment by HeatherSoron** [comment]
Haha, at first glance this looks awesome :D! I'll read this over, make sure it's safe (since it runs directly on the user's computer), and also see if it works on Linux with Node v4.2.6.

Might take a bit for me to do the pull request review, but if this is what I think it is, it'll be quite awesome to have :).

**Comment by HeatherSoron** [comment]
Wait, here's a thought: all the standard Node functions are still available inside this, aren't they? So, someone could write a script which calls require(), or which uses child_process.exec, couldn't they...?

Okay, even if there's nothing wrong with _this_ code, I'll still have to put a big bolded warning to let people know what risks there are.

**Comment by memetech** [comment]
Hmmm, that leads in to a whole different kind of phishing. My intent was that people use this to test their own scripts, not get them from other people and test them here. But I could see someone borrowing a library and wanting to test it with their own scripts. Probably a warning is a good idea I agree!

if you do not want to keep this framework alongside the sample scripts, that's ok too. I hadn't thought about this aspect.

**Comment by HeatherSoron** [comment]
This is a pretty useful tool, so I'm inclined to accept it - just with a lot of warnings for the user. Alternatively, I might see if there's some way in Node to spawn a new, bare-bones V8 context, with none of the standard dangerous things available. I mean, V8 itself is pretty safe, it's just that Node adds a lot of useful (and dangerous) extras.

File: testing/fwtests/dbtest1.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+function(context, args) {
+
+    var k = "akey";
+    var o = { obj:"obj" };
+
+    var i = #db.i({ k: k, o: o });
+
+    var o2 = #db.f({ k: k }).array();
+    console.log(o2);
```
Comment (HeatherSoron) [inline testing/fwtests/dbtest1.js:9]: Doesn't actually work in hackmud, FYI. (`console` doesn't exist, so there's no `console.log`)

Comment (memetech) [reply testing/fwtests/dbtest1.js:9]: True, this fwtests directory is supposed to only test the test framework code, but not run in hackmud.

**Comment by memetech** [comment]
Ok, here's the big problem you identified, when somebody tried to plug in a hackmud script consisting of require('fs') (which I assume would not run in hackmud!):

=======Running test script: ./fwtests/fsphishtest=======
===     Loading script with adaptations...
===     Connecting to MongoDB...
===     Running the test code:
[ 'Stats',
  'F_OK',
  'R_OK',
  'W_OK',
  'X_OK',
  'access',
  'accessSync',
  'exists',
  'existsSync',
  'readFile',
  'readFileSync',
  'close',
  'closeSync',
  'open',
  'openSync',
  'read',
  'readSync',
  'write',
  'writeSync',
  'rename',
  'renameSync',
  'truncate',
  'truncateSync',
  'ftruncate',
  'ftruncateSync',
  'rmdir',
  'rmdirSync',
  'fdatasync',
  'fdatasyncSync',
  'fsync',
  'fsyncSync',
  'mkdir',
  'mkdirSync',
  'readdir',
  'readdirSync',
  'fstat',
  'lstat',
  'stat',
  'fstatSync',
  'lstatSync',
  'statSync',
  'readlink',
  'readlinkSync',
  'symlink',
  'symlinkSync',
  'link',
  'linkSync',
  'unlink',
  'unlinkSync',
  'fchmod',
  'fchmodSync',
  'chmod',
  'chmodSync',
  'fchown',
  'fchownSync',
  'chown',
  'chownSync',
  '_toUnixTimestamp',
  'utimes',
  'utimesSync',
  'futimes',
  'futimesSync',
  'writeFile',
  'writeFileSync',
  'appendFile',
  'appendFileSync',
  'watch',
  'watchFile',
  'unwatchFile',
  'realpathSync',
  'realpath',
  'createReadStream',
  'ReadStream',
  'FileReadStream',
  'createWriteStream',
  'WriteStream',
  'FileWriteStream' ]
=OK     Clean exit from test code!
===     Cleaning up MongoDB

**Comment by HeatherSoron** [comment]
Alright, the code you've submitted all seems to be safe and non-malicious. Doesn't work in Node v4.2.6, due to really-need breaking. That's not a fatal issue, though; I can probably rewrite the essential bit myself (I've done a `require` hack before).

Now I just need to get a **nice big giant warning** about not running other people's code in this test framework, at least until there's a robust sandbox that blocks `require` and such ;). I can write this if needed.

**Comment by HeatherSoron** [comment]
By the way, here's a pointer towards a safer way to run JS code within Node: https://nodejs.org/api/vm.html

There are methods on that which allow you to create a VM that doesn't have `require`, but instead has a caller-defined set of global variables. That would let the test framework pass in specific data (script references, DB functions, etc.), without exposing `require` and other things which are interesting/scary/non-hackmud.

**Comment by Spartan322** [comment]
Yeah, using node's built in vm strips the need for really-need and further sandboxes the code being running, keeping it from possibly blowing up the running node process. (If you pull the code by reading it to a string instead of using require)

**Pull request closed without merging**